libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp

```diff
@@ -91,3 +91,128 @@ void AP_BattMonitor_Backend::update_resistance_estimate()
     _state.voltage_resting_estimate = _state.voltage + _state.current_amps * _state.resistance;
 }
+
+float AP_BattMonitor_Backend::voltage_resting_estimate() const
+{
+    // resting voltage should always be greater than or equal to the raw voltage
+    return MAX(_state.voltage, _state.voltage_resting_estimate);
+}
+
+AP_BattMonitor::BatteryFailsafe AP_BattMonitor_Backend::update_failsafes(void)
+{
+    const uint32_t now = AP_HAL::millis();
+
+    bool low_voltage, low_capacity, critical_voltage, critical_capacity;
+    check_failsafe_types(low_voltage, low_capacity, critical_voltage, critical_capacity);
+
+    if (critical_voltage) {
+        // this is the first time our voltage has dropped below minimum so start timer
+        if (_state.critical_voltage_start_ms == 0) {
+            _state.critical_voltage_start_ms = now;
+        } else if (_params._low_voltage_timeout > 0 &&
+                   now - _state.critical_voltage_start_ms > uint32_t(_params._low_voltage_timeout)*1000U) {
+            return AP_BattMonitor::BatteryFailsafe_Critical;
+        }
+    } else {
+        // acceptable voltage so reset timer
+        _state.critical_voltage_start_ms = 0;
+    }
+
+    if (critical_capacity) {
+        return AP_BattMonitor::BatteryFailsafe_Critical;
+    }
+
+    if (low_voltage) {
+        // this is the first time our voltage has dropped below minimum so start timer
+        if (_state.low_voltage_start_ms == 0) {
+            _state.low_voltage_start_ms = now;
+        } else if (_params._low_voltage_timeout > 0 &&
+                   now - _state.low_voltage_start_ms > uint32_t(_params._low_voltage_timeout)*1000U) {
+            return AP_BattMonitor::BatteryFailsafe_Low;
+        }
+    } else {
+        // acceptable voltage so reset timer
+        _state.low_voltage_start_ms = 0;
+    }
+
+    if (low_capacity) {
+        return AP_BattMonitor::BatteryFailsafe_Low;
+    }
+
+    // if we've gotten this far then battery is ok
+    return AP_BattMonitor::BatteryFailsafe_None;
+}
+
+bool update_check(size_t buflen, char * buffer, bool condition, const char *message) {
+    if (condition) {
+        strncpy(buffer, message, buflen);
+        return false;
+    }
+    return true;
+}
+
+bool AP_BattMonitor_Backend::arming_checks(size_t buflen, char * buffer) const
+{
+    bool result = true;
+
+    bool low_voltage, low_capacity, critical_voltage, critical_capacity;
+    check_failsafe_types(low_voltage, low_capacity, critical_voltage, critical_capacity);
+
+    bool below_arming_voltage = is_positive(_params._arming_minimum_voltage) &&
+                                (_state.voltage < _params._arming_minimum_voltage);
+    bool below_arming_capacity = (_params._arming_minimum_capacity > 0) &&
+                                 ((_params._pack_capacity - _state.consumed_mah) < _params._arming_minimum_capacity);
+
+    result = result && update_check(buflen, buffer, low_voltage,  "low voltage failsafe");
```

Comment:
I'm uneasy with the interdependence of the ordering here and in check_failsafe_types`.  In particular, I think you might be able to be both low-voltage and critical-voltage at the same time.  The ordering here would mean you emit low-voltage but not critical-voltage.

Reply:
I swapped the arguments so we always print all of the failures. I wasn't to worried about it since I was assuming one would always resolve all battery errors, but it's an easy rearrangement.

Reply:
Swapping the order doesn't actually help so I ended up reverting that. Unless I can send the status text from the library it's impossible to guarantee sending the highest priority message, and the order isn't fundamentally clear. You will only get in trouble here if you disable the arming check, in which case the vehicle will still failsafe on you as soon as you fly (and you killed the check) so I'm inclined to live with this.